File: prisma/refs/java/design-word-search-data-structure.java
class WordDictionary {
    private static class Node {
        Node[] children = new Node[26];
        boolean end;
    }

    private final Node root = new Node();

    public WordDictionary() {}

    public void addWord(String word) {
        Node node = root;
        for (char c : word.toCharArray()) {
            if (node.children[c - 'a'] == null) node.children[c - 'a'] = new Node();
            node = node.children[c - 'a'];
        }
        node.end = true;
    }

    public boolean search(String word) {
        return dfs(root, word, 0);
    }

    private boolean dfs(Node node, String word, int i) {
        if (i == word.length()) return node.end;
        char c = word.charAt(i);
        if (c == '.') {
            for (Node child : node.children)
                if (child != null && dfs(child, word, i + 1)) return true;
            return false;
        }
        Node next = node.children[c - 'a'];
        return next != null && dfs(next, word, i + 1);
    }
}

File: prisma/refs/py/design-word-search-data-structure.py
class WordDictionary:
    def __init__(self):
        self.root = {}

    def addWord(self, word):
        node = self.root
        for c in word:
            node = node.setdefault(c, {})
        node["$"] = True

    def search(self, word):
        def dfs(node, i):
            if i == len(word):
                return "$" in node
            c = word[i]
            if c == ".":
                return any(dfs(child, i + 1) for k, child in node.items() if k != "$")
            return c in node and dfs(node[c], i + 1)

        return dfs(self.root, 0)
